### PR TITLE
Expose citation data in HTML

### PR DESCRIPTION
**Pull request by mattdahl**
This PR makes a series of changes to the view logic and the template HTML in order to render the new citation depth information in helpful places. Summary of changes:

1. First, in ffad2ed, I changed how we calculate a cluster's unique citation count because things seemed out of sync for some reason on my local machine. I think my changes should be a safer way to calculate that number.
2. Then, in 14c18f2, I created a new property on the `OpinionCluster` object called `authorities_with_data`, which basically wraps the existing `authorities` property with some additional citation depth data. This is then passed into the appropriate templates and rendered.
3. In 8d2aaff, I do more or less the same thing, creating new `children` and `children_with_data` properties as well. To create `children`, I just transferred the existing Solr lookup logic that was used in `cl/opinion_page/views.py`. And as above, `children_with_data` appends the additional citation data to that returned dictionary. `children_with_data` is then also passed to the appropriate templates and rendered.
4. Finally, in 546c236 I also render the depth data on search result pages, but only if the search querystring contains a single `cites:(X)` term. I used regex to calculate this state (there may be a better way?).

For viewing the changes visually, these are the pages with new data being rendered on them:
- The main opinion page (with "(x times)" strings in the "cited by" and "authorities" boxes in the left sidebar)
- The "view all authorities" page for an opinion (with "x references to" strings at the beginning of each case name)
- The "view all citing opinions" page for an opinion/search results page with a `cites:(X)` term (with "x references to" strings underneath each case name)

### CODE REVIEW

**Comment by mlissner** [comment]
I have a bit of anxiety around the performance of these changes. I am going to run this branch a bit tomorrow and see if I can watch the queries that are run. Definitely some tricky stuff here.

**Comment by mattdahl** [comment]
Certainly, that makes sense

File: cl/search/views.py [hunk 1]
```diff
@@ -174,6 +176,26 @@ def do_search(
...
+        except OpinionCluster.DoesNotExist:
+            pass
+        else:
+            for result in paged_results.object_list:
+                result[
+                    "references_count"
+                ] = get_citation_depth_for_cluster_pks(
+                    citing_cluster_pk=result["id"],
+                    cited_cluster_pk=cited_cluster.pk,
+                )
```
Comment (mlissner) [inline cl/search/views.py]: This looks pretty good, but can we move it to its own function, perhaps in search_utils.py? And can we add a note, perhaps, about why we only do it for single matches?

File: cl/search/models.py [hunk 2]
```diff
@@ -2158,6 +2161,80 @@ def has_private_authority(self):
...
+                "sub_opinions": authority.sub_opinions,
+                # Number of other opinions that cite this authority
+                "unique_citation_count": authority.citation_count,
+                # Number of citations from this cluster to this authority
+                "this_cluster_citation_count": get_citation_depth_for_cluster_pks(
+                    citing_cluster_pk=self.pk, cited_cluster_pk=authority.pk,
+                ),
+            }
+            for authority in self.authorities
+        ]
```
Comment (mlissner) [inline cl/search/models.py]: I'm guessing you tried this, but can't you cast a QuerySet to a list and then just add this attribute to each item, then sort the list? I'm surprised that's not possible, I guess, and I suspect it'd be a lot tidier?

File: cl/opinion_page/views.py [hunk 3]
```diff
@@ -387,8 +368,10 @@ def view_opinion(request, pk, _):
             "favorite_form": favorite_form,
             "get_string": get_string,
             "private": cluster.blocked,
-            "citing_clusters": citing_clusters,
-            "top_authorities": cluster.authorities[:5],
+            "top_children": children[:5],
+            "children_count": len(children),
```
Comment (mlissner) [inline cl/opinion_page/views.py]: Just checking: You tested that `len()` works like this? I'm a bit surprised, but I've been away from the Sunburnt APIs for a bit.

File: cl/opinion_page/views.py [hunk 4]
```diff
@@ -387,8 +368,10 @@ def view_opinion(request, pk, _):
             "favorite_form": favorite_form,
             "get_string": get_string,
             "private": cluster.blocked,
-            "citing_clusters": citing_clusters,
-            "top_authorities": cluster.authorities[:5],
+            "top_children": children[:5],
```
Comment (mlissner) [inline cl/opinion_page/views.py]: Does slicing the array like this take the full results and slice them or does it only query Solr for five results? The difference could be significant.

File: cl/opinion_page/views.py [hunk 5]
```diff
@@ -354,28 +354,9 @@ def view_opinion(request, pk, _):
...
-            "rows": 5,
-            "start": 0,
-            "sort": "citeCount desc",
-            "caller": "view_opinion",
-        }
-        citing_clusters = conn.raw_query(**q).execute()
+        children = cluster.children_with_data
     else:
-        citing_clusters = None
+        children = None
```
Comment (mlissner) [inline cl/opinion_page/views.py]: This sounds dumb, but can we improve this for bots? What would happen if we cached the values the humans got the first time and then later showed those values to the bots? So bots can't generate data, but if humans already have, then the bots can access it? If we cache to the DB, it'll be fast and take up negligible space? Perhaps a tweak for later. It'd be good for SEO (something I don't want to care about, but sadly do).

File: cl/opinion_page/templates/view_opinion.html [hunk 6]
```diff
@@ -76,24 +76,25 @@ <h3 class="inline"><span>Your Notes</span></h3>
...
+                               title="Subscribe to a feed of citations to this case."></i>
+                        </a>
+                    </span>
+                </h3>
+                <p class="bottom">This case has been cited by these opinions:</p>
+                <ul>
+                    {% for child in top_children %}
+                        <li>
+                            <a href="{{ child.absolute_url }}?{{ request.META.QUERY_STRING }}">{{ child.caseName|safe|truncatewords:12|v_wrapper }} ({{ child.dateFiled|date:"Y" }})</a>
+                            ({{ child.citations_count }} time{{ child.citations_count|pluralize }})
```
Comment (mlissner) [inline cl/opinion_page/templates/view_opinion.html]: Would the `gray` CSS class on a `<span>` look better here? Or maybe the `small` class?

File: cl/opinion_page/templates/view_opinion.html [hunk 7]
```diff
@@ -104,16 +105,17 @@ <h4>
...
+                <h3><span>Authorities ({{ authorities_count }})</span></h3>
                 <p class="bottom">This opinion cites:</p>
                 <ul>
                     {% for authority in top_authorities %}
                         <li>
-                            <a href="{{ authority.get_absolute_url }}?{{ request.META.QUERY_STRING }}">
+                            <a href="{{ authority.absolute_url }}?{{ request.META.QUERY_STRING }}">
                                 {{ authority.caption|safe|truncatewords:10|v_wrapper }}
                             </a>
+                            ({{ authority.this_cluster_citation_count }} time{{ authority.this_cluster_citation_count|pluralize }})
```
Comment (mlissner) [inline cl/opinion_page/templates/view_opinion.html]: Same q as above. Would `gray` and/or `small` be good here?

File: cl/citations/utils.py [hunk 8]
```diff
@@ -18,3 +22,22 @@ def map_reporter_db_cite_type(citation_type):
         "scotus_early": Citation.SCOTUS_EARLY,
     }
     return citation_map.get(citation_type)
+
+
+def get_citation_depth_for_cluster_pks(citing_cluster_pk, cited_cluster_pk):
```
[the remark this answers is not in the document]
Comment (mattdahl) [reply cl/citations/utils.py]: Agree about the "between," but I think the "by_pks" is actually too verbose? The parameter names make it obvious that pks must be passed. My changes are in 87087af.

File: cl/citations/utils.py [hunk 9]
```diff
@@ -18,3 +22,22 @@ def map_reporter_db_cite_type(citation_type):
...
+    """OpinionsCited objects exist as relationships between Opinion objects,
+    but we often want access to citation depth information between
+    OpinionCluster objects. This helper method assists in doing the necessary
+    DB lookup.
+
+    :param citing_cluster_pk: The primary key of the citing OpinionCluster
+    :param citing_cluster_pk: The primary key of the cited OpinionCluster
```
[the remark this answers is not in the document]
Comment (mattdahl) [reply cl/citations/utils.py]: Fixed in 87087af

File: cl/search/models.py [hunk 10]
```diff
@@ -2158,6 +2161,80 @@ def has_private_authority(self):
...
+        conn = sunburnt.SolrInterface(settings.SOLR_OPINION_URL, mode="r")
+        q = {
+            "q": "cites:({ids})".format(
+                ids=" OR ".join(
+                    [
+                        str(pk)
+                        for pk in (
+                            self.sub_opinions.values_list("pk", flat=True)
+                        )
+                    ]
```
[the remark this answers is not in the document]
Comment (mattdahl) [reply cl/search/models.py]: Changes reverted in f85e738

File: cl/search/models.py [hunk 11]
```diff
@@ -2158,6 +2161,80 @@ def has_private_authority(self):
...
+                )
+            ),
+            "start": 0,
+            "caller": "view_opinion",
```
[the remark this answers is not in the document]
Comment (mattdahl) [reply cl/search/models.py]: Changes reverted in f85e738

File: cl/search/models.py [hunk 12]
```diff
@@ -2158,6 +2161,80 @@ def has_private_authority(self):
...
+    def children_with_data(self):
+        """Returns a list of dictionaries containing useful data on every
+        child, for eventual injection into a view template. Because
+        self.children is already a list of dictionaries (a Solr result), it's
+        easy to calculate and append an extra data field (relating to citation
+        counts) on-the-fly.
+        The returned list is then sorted by that citation count field.
+        """
+        children_with_data = self.children
+        for child in children_with_data:
```
[the remark this answers is not in the document]
Comment (mattdahl) [reply cl/search/models.py]: Changes reverted in f85e738

File: cl/opinion_page/views.py (hunk 3 above)
Comment (mattdahl) [reply cl/opinion_page/views.py]: Changes reverted in f85e738

File: cl/opinion_page/views.py (hunk 4 above)
Comment (mattdahl) [reply cl/opinion_page/views.py]: Changes reverted in f85e738

File: cl/opinion_page/views.py (hunk 5 above)
Comment (mattdahl) [reply cl/opinion_page/views.py]: Relevant changes reverted in f85e738, but perhaps this comment should be preserved for further discussion

File: cl/opinion_page/templates/view_opinion.html (hunk 6 above)
Comment (mattdahl) [reply cl/opinion_page/templates/view_opinion.html]: Changes reverted in f85e738

File: cl/search/models.py (hunk 2 above)
Comment (mattdahl) [reply cl/search/models.py]: Ah, for some reason I thought it was not possible to add a new attribute to an object that had already been instantiated, so I thought I needed to wrap each `OpinionCluster` object in a dictionary! But thank god for dynamic typing. This is much simplified now: 2ffe8c7

File: cl/search/views.py (hunk 1 above)
Comment (mattdahl) [reply cl/search/views.py]: Fixed in b3669b4, though I'm not entirely satisfied by my refactor. Couldn't think of a good name for the helper function -- what do you think?

File: cl/opinion_page/templates/view_opinion.html (hunk 7 above)
Comment (mattdahl) [reply cl/opinion_page/templates/view_opinion.html]: I think both improve the aesthetic. Fixed in 6052de5

File: cl/citations/tasks.py [hunk 13]
```diff
@@ -144,6 +137,14 @@ def find_citations_for_opinion_by_pks(self, opinion_pks, index=True):
...
             )
 
+            # Recalculate the citation counts for the matched clusters
+            for matched_opinion in grouped_matches:
+                count = OpinionsCited.objects.filter(
+                    cited_opinion__cluster__pk=matched_opinion.cluster.pk,
+                ).count()
+                matched_opinion.cluster.citation_count = count
+                matched_opinion.cluster.save(index=index)
+
```
[the remark this answers is not in the document]
Comment (mattdahl) [reply cl/citations/tasks.py]: Reverted in 6a88f35; will take up in new PR

File: cl/search/views.py (hunk 1 above)
Comment (mlissner) [reply cl/search/views.py]: It looks about right. The name can probably be simplified to something like `add_depth_counts`?

File: cl/search/models.py (hunk 2 above)
Comment (mlissner) [reply cl/search/models.py]: Calling this `cluster_references` is a little confusing, isn't it? Why not just call it `depth` or something?

File: cl/opinion_page/views.py (hunk 5 above)
Comment (mlissner) [reply cl/opinion_page/views.py]: Moved to: https://github.com/freelawproject/courtlistener/issues/1197

File: cl/search/views.py (hunk 1 above)
Comment (mattdahl) [reply cl/search/views.py]: 👍

File: cl/search/models.py (hunk 2 above)
Comment (mattdahl) [reply cl/search/models.py]: 👍

**Pull request closed without merging**